onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc

```diff
@@ -0,0 +1,441 @@
+    trt_builder->setMaxWorkspaceSize(kMaxWorkSpaceSize);
+    auto trt_engine = unique_pointer<nvinfer1::ICudaEngine>(trt_builder->buildCudaEngine(*trt_network.get()));
+    ORT_ENFORCE(trt_engine != nullptr);
+
+    // Build TensorRT context
+    auto trt_context = unique_pointer<nvinfer1::IExecutionContext>(trt_engine->createExecutionContext());
+    ORT_ENFORCE(trt_context != nullptr);
+
+    // Get input shape and binding index
+    int num_inputs = trt_network->getNbInputs();
+    input_indexes.resize(num_inputs);
+    input_dim_sizes.resize(num_inputs);
+    for (int i = 0; i < num_inputs; ++i) {
+      const std::string& name = trt_network->getInput(i)->getName();
+      size_t bindingIndex = trt_engine->getBindingIndex(name.c_str());
+      nvinfer1::Dims dimensions = trt_engine->getBindingDimensions(static_cast<int>(bindingIndex));
+      auto iter = input_map.find(name);
+      if (iter != input_map.end()) {
+        input_indexes[bindingIndex] = iter->second;
+      }
+      size_t dim_size = 1;
+      for (int j = 0, end = dimensions.nbDims; j < end; ++j) {
+        dim_size *= dimensions.d[j];
+      }
+      input_dim_sizes[bindingIndex] = dim_size;
+    }
+
+    // Get output shape and binding index
+    int num_outputs = trt_network->getNbOutputs();
+    output_indexes.resize(num_outputs);
+    output_dim_sizes.resize(num_outputs);
+    output_shapes.resize(num_outputs);
+    for (int i = 0; i < num_outputs; ++i) {
+      const std::string& name = trt_network->getOutput(i)->getName();
+      size_t bindingIndex = trt_engine->getBindingIndex(name.c_str());
+      nvinfer1::Dims dimensions = trt_engine->getBindingDimensions(static_cast<int>(bindingIndex));
+      bindingIndex -= num_inputs;
+      auto iter = output_map.find(name);
+      if (iter != output_map.end()) {
+        output_indexes[bindingIndex] = iter->second;
+      }
+      size_t dim_size = 1;
+      for (int j = 0, end = dimensions.nbDims; j < end; ++j) {
+        output_shapes[bindingIndex].push_back(dimensions.d[j]);
+        dim_size *= dimensions.d[j];
+      }
+      output_dim_sizes[bindingIndex] = dim_size;
+    }
+
+    ORT_ENFORCE(trt_engine->getNbBindings() == (num_inputs + num_outputs));
+
+    // Save engine, context and input/output info to map
+    parsers_.emplace(fused_node->Name(), std::move(trt_parser));
+    engines_.emplace(fused_node->Name(), std::move(trt_engine));
+    contexts_.emplace(fused_node->Name(), std::move(trt_context));
+    input_info_[fused_node->Name()].push_back(input_indexes);
+    input_info_[fused_node->Name()].push_back(input_dim_sizes);
+    output_info_[fused_node->Name()].push_back(output_indexes);
+    output_info_[fused_node->Name()].push_back(output_dim_sizes);
+    output_shapes_[fused_node->Name()] = output_shapes;
+
+    // Create function state
+    NodeComputeInfo compute_info;
+    compute_info.create_state_func = [=](ComputeContext* context, FunctionState* state) {
```

Comment:
@jywu-msft @stevenlix The create function is designed to work across DLL boundaries and this kind of default capture is going to prevent from that happening if (in the future) we decide to move this functionality to a separate DLL. This breaks the design principle behind the 3 functions. If we're not going to fix it in this PR, can we mark a TODO?

Also, please avoid default capture.

Reply:
I added a TODO comment and will remove the default capture soon.